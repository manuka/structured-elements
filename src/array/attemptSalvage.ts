import type { StructuredElements } from "@"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const attemptArraySalvage: StructuredElements.Functions.AttemptSalvage<
  `array`
> = <Registry extends StructuredElements.BaseRegistry, Element>({
  failures,
  subject,
  validElements,
}: {
  failures: StructuredElements.Failure[]
  subject: unknown
  validElements?: Element[]
}): Element[] | undefined => {
  if (failures.length === 0) {
    return subject as Element[]
  }

  if (validElements) {
    return validElements
  }

  return []
}
