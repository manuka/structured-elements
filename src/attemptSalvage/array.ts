import type { StructuredElements } from "@"

export const attemptSalvageArray: StructuredElements.Functions.AttemptSalvage<
  `array`
> = <Registry extends StructuredElements.BaseRegistry, Element>({
  api: _,
  failures,
  subject,
  validElements,
}: {
  api: StructuredElements.API<Registry>
  failures: StructuredElements.Failure[]
  subject: unknown
  validElements?: Element[]
}): Element[] | undefined => {
  if (failures.length === 0) {
    return subject as Element[]
  }

  if (validElements) {
    return validElements
  }

  return []
}
