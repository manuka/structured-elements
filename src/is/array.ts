export const isArray = <T>(subject: unknown): subject is T[] => {
  return Array.isArray(subject)
}
