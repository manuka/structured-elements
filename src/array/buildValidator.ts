import { attemptArraySalvage } from "@/array/attemptSalvage"
import { buildStructuredElementsResult } from "@/buildResult"
import { isArray } from "@/isArray"
import { validateElement } from "@/validateElement"

import type { StructuredElements } from "@"

const structure = `array` as const

export const buildArrayValidator = <
  Registry extends StructuredElements.BaseRegistry,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Element,
>({
  api,
  expectation,
}: {
  api: StructuredElements.API<Registry>
  expectation: StructuredElements.Expectation<Registry, Element>
}): StructuredElements.Validator<Element, `array`> => {
  const validator: StructuredElements.Validator<Element, `array`> = {
    getFailures: (subject, name) => {
      return validator.validate(subject, name).failures
    },

    getSalvage: (subject, name, attemptSalvage) => {
      return validator.validate(subject, name, attemptSalvage).salvage
    },

    isValid: (subject, name): subject is Element[] => {
      return validator.validate(subject, name).valid
    },

    validate: (subject, name, attemptSalvage = attemptArraySalvage) => {
      expectation
      if (!isArray(subject)) {
        return buildStructuredElementsResult({
          failures: [
            {
              element: subject,
              expectation,
              key: `(subject)`,
              name,
              reason: `${name} expected to be an array, but isn't. It has typeof ${typeof subject}`,
              subject,
            },
          ],
          subject,
        })
      }

      const cached = api.privateFunctions.getCachedResult({
        expectation,
        structure,
        subject,
      })

      if (cached) {
        return cached
      }

      const failures: StructuredElements.Failure[] = []
      const validElements: Element[] = []

      subject.forEach((element, index) => {
        const elementFailures = validateElement({
          api,
          element,
          expectation,
          key: index.toString(),
          subject,
          subjectName: name,
        })

        if (elementFailures.length > 0) {
          failures.push(...elementFailures)
        } else {
          validElements.push(element as Element)
        }
      })

      const salvage = attemptSalvage({
        api,
        failures,
        name,
        subject,
        validElements,
      })

      const result = buildStructuredElementsResult({
        failures,
        salvage,
        subject,
      })

      api.privateFunctions.cacheResult({
        expectation,
        result,
        structure,
      })

      return result
    },
  }

  return validator
}
