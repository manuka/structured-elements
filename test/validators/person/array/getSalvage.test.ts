import { setupValidatorPersonArrayTest } from "&/validators/person/array/common"
import { testNestedScenarios } from "test-nested-scenarios"

describe(`validator('Person', 'array').getSalvage(\n  {`, () => {
  const setup = setupValidatorPersonArrayTest({
    expectation: ({ subject, validator }) => {
      it(`},\n  'Person'\n)`, () => {
        expect(validator.getSalvage(subject, `Person`)).toMatchSnapshot()
      })
    },
  })
  testNestedScenarios(setup)
})
