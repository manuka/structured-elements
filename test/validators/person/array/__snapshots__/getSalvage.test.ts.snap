// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: (Invalid), person2: (Invalid), },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: (Invalid), person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: (Invalid), person2: undefined, },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: (Valid), person2: (Invalid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: (Valid), person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: (Valid), person2: undefined, },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: undefined, person2: (Invalid), },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: undefined, person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Invalid), person1: undefined, person2: undefined, },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: (Invalid), person2: (Invalid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: (Invalid), person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: (Invalid), person2: undefined, },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: (Valid), person2: (Invalid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: (Valid), person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: (Valid), person2: undefined, },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: undefined, person2: (Invalid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: undefined, person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: (Valid), person1: undefined, person2: undefined, },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-0",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: (Invalid), person2: (Invalid), },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: (Invalid), person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: (Invalid), person2: undefined, },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: (Valid), person2: (Invalid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: (Valid), person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: (Valid), person2: undefined, },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-1",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: undefined, person2: (Invalid), },
  'Person'
) 1`] = `[]`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: undefined, person2: (Valid), },
  'Person'
) 1`] = `
[
  {
    "id": "Valid Person.id-2",
    "inventory": [
      {
        "id": "Valid Thing.id",
        "name": "Valid Thing.name",
        "parts": {},
        "type": "gadget",
        "weight": 1,
      },
    ],
    "name": "Valid Person.name",
  },
]
`;

exports[`validator('Person', 'array').getSalvage(
  { person0: undefined, person1: undefined, person2: undefined, },
  'Person'
) 1`] = `[]`;
